File: 2007-find-original-array-from-doubled-array/2007-find-original-array-from-doubled-array.java
class Solution {
    public int[] findOriginalArray(int[] changed) {
        
        if(changed.length % 2 != 0)
			return new int[0];
		
		int[] ans = new int[changed.length / 2];
		
		HashMap<Integer,Integer> map = new HashMap<>();
		
		Arrays.sort(changed);
		
		for(int i = 0; i < changed.length; i++)
			map.put(changed[i], map.getOrDefault(changed[i], 0)+1);
		
		int indx = 0;
		
		for(int i = 0; i < changed.length; i++)
		{
			if(map.get(changed[i])<1) // keep track of the used items
				continue;
			else if(map.getOrDefault(changed[i]*2, 0) < 1) // Check if the double Exists
				return new int[0];
			else {
				map.put(changed[i], map.get(changed[i]) - 1);
				map.put(changed[i]*2, map.get(changed[i]*2) - 1);
				ans[indx] = changed[i];
				indx++;
			}
			
		}
		
		return ans;
	}
}

File: 2007-find-original-array-from-doubled-array/2007-find-original-array-from-doubled-array.py
class Solution:
    def findOriginalArray(self, changed: List[int]) -> List[int]:
        if len(changed) % 2 == 1:
            return []
        changed.sort()
        dic = dict()
        for i in changed:
            if dic.get(i):
                dic[i] += 1
            else:
                dic[i] = 1
        ans = []
        for i in changed:
            if i == 0:
                if dic.get(i) >= 2:
                    dic[i] -= 2
                    ans.append(i)
            elif i > 0 and dic.get(i) and dic.get(i * 2):
                ans.append(i)
                dic[i * 2] -= 1
                dic[i] -= 1
        if len(ans) == len(changed) // 2:
            return ans
        return []